### ISSUE TITLE
`isapprox` is not defined for arguments for which `==` is defined

### ISSUE THREAD

**Issue by wsshin** [JuliaLang/julia#19035]
Both `==` and `≈` can be used for comparing two arrays:

```
julia> [1, 2, 3] == [1, 2, 3]
true

julia> [1, 2, 3] ≈ [1, 2, 3]
true
```

On the other hand, `==` works for two arrays of arrays, whereas `≈` does not:

```
julia> [[1, 2], [3, 4]] == [[1, 2], [3, 4]]
true

julia> [[1, 2], [3, 4]] ≈ [[1, 2], [3, 4]]
ERROR: MethodError: no method matching isapprox(::Array{Array{Int64,1},1}, ::Array{Array{Int64,1},1})
```

Is this an expected behavior?

Here is the version info:

```
julia> versioninfo()
Julia Version 0.5.1-pre+2
Commit f0d40ec* (2016-09-20 03:34 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin16.0.0)
  CPU: Intel(R) Core(TM)2 Duo CPU     T9900  @ 3.06GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Penryn)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, penryn)
```

**Comment by TotalVerb** [comment]
Since `norm` is recursive (in a well-defined way), this is a bug and not an expected behaviour.

**Comment by wsshin** [comment]
Thanks!  Hope some Julia members to label this issue as a bug.

**Comment by martinholters** [comment]
The issue title is quite broad, but this should still be limited to the numeric domain, right?
At first glance it seems to make sense that if `a == b` returns `true`, then so should `a ≈ b`, but introducing a `isapprox(x, y) = x == y` fallback is not what this is heading to, is it?

**Comment by TotalVerb** [comment]
I think that any type for which norm and subtraction are defined should get an isapprox.

**Comment by StefanKarpinski** [comment]
If this gets done in time for 0.6, totally fine, but it's not release-critical.

**Comment by akaysh** [comment]
Hi @StefanKarpinski 
Can this be worked on by a beginner? I want to take up this issue.

**Comment by wsshin** [comment]
See my PR https://github.com/JuliaLang/julia/pull/19637 that solves this issue.  

The goal was to support `isapprox()` for nested arrays with arbitrary nesting levels.  For example, I wanted `isapprox()` to handle something like `a = [[1,2, [3,4]], 5.0, [6im, [7.0, 8.0]]]`.  Because `norm(a)` is defined for such `a`, `isapprox()` should be able to handle `a` as @TotalVerb pointed out.

The main difficulties in generalizing `isapprox()` to support nested arrays were as follows:
- The previous implementation assumed that the arguments are of `AbstractArary{T<:Number}` type, but to support nested arrays, I had to change this to `AbstractArray`.
- This change made it difficult to assign the default value to `rtol`.  Previously the default value was `Base.rtoldefault(T,S)`, but this was no longer possible because nested arrays could have inhomogeneous leaf element types (e.g., in the above `a`, all the `Int64`, `Float64`, and `Complex128` types were used as leaf elements).

So, I implemented `promote_leaf_eltypes()`, which takes a nested array, examine its all leaf types, and find a promoted type from them.  Then, for the two nested arrays `x` and `y` given to `isapprox()`, I assigned `Base.rtoldefault(promote_leaf_eltypes(x), promote_leaf_eltypes(y))` as the default value of `rtol`.

The code seems working well.  In the PR, I also added tests for the new `isapprox()` and `promote_leaf_eltypes()`.
